### PR TITLE
feat: a middleware version with tests to storage explorer service

### PR DESCRIPTION
**Pull request by assafchamoy**
| Question                | Answer                                                                          |
| ---------------- | -------------------------------------------------------------------------- |
| Bug fix         | ✖                                                                        |
| New feature     | ✔                                                                     |
| Breaking change | ✔                                                                       |
| Deprecations    | ✖                                                                        |
| Documentation   | ✖                                                                        |
| Tests added     | ✔                                                                     |
| Chore            | ✖                                                                       |

Further  information:
A working middleware version of https://github.com/MapColonies/storage-explorer.
Tests includes integration and unit.
express-test server to use the middlware on, and test the endpoints.

### CODE REVIEW

File: src/common/interfaces.ts [hunk 1]
```diff
@@ -0,0 +1,12 @@
+import { ReadStream } from 'fs';
```
Comment (alebinson) [inline src/common/interfaces.ts:1]: Worth to fix Lint errors

File: src/common/utilities/LoggersHandler.ts [hunk 2]
```diff
@@ -0,0 +1,19 @@
+export default class LoggersHandler {
+  public constructor(private readonly logger: Record<any, any>) {}
```
Comment (alebinson) [inline src/common/utilities/LoggersHandler.ts]: Lint errors

File: src/common/utilities/LoggersHandler.ts [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+
+  public info(...payload: string[]): void {
+    if (Object.prototype.hasOwnProperty.call(this.logger, 'info') && this.logger.info instanceof Function) {
```
Comment (alebinson) [inline src/common/utilities/LoggersHandler.ts]: think, this kind of check might be good enough
**if (typeof this.logger.info === "function") { 
    // safe to use the function
}**
Even might be extracted to separate function

File: src/storageExplorer/middlewares/getStorageExplorerMiddleware.ts [hunk 4]
```diff
@@ -0,0 +1,13 @@
+import { Router } from 'express';
+import { ImountDirObj } from '../../common/interfaces';
+import LoggersHandler from '../../common/utilities/LoggersHandler';
+import { StorageExplorerController } from '../controllers/storageExplorer.controller';
+import { explorerRoutes } from '../routes/storageExplorer.router';
+
+export default function getStorageExplorerMiddleware(mountDirs: ImountDirObj[], logger: Record<any, any>): Router {
```
Comment (alebinson) [inline src/storageExplorer/middlewares/getStorageExplorerMiddleware.ts]: Lint errors

File: src/storageExplorer/models/file.model.ts [hunk 5]
```diff
@@ -0,0 +1,12 @@
+// Part of 'chonky' FileData interface https://github.com/TimboKZ/Chonky/blob/2.x/packages/chonky/src/types/file.types.ts
```
Comment (alebinson) [inline src/storageExplorer/models/file.model.ts]: Think chonky remainder isn't relevant. You are right, it's inspired by chonky's type but not more than that

File: tests/unit/storageExplorer/models/DirOperations.spec.ts [hunk 6]
```diff
@@ -0,0 +1,120 @@
+/* eslint-disable @typescript-eslint/no-misused-promises */
+import { Dirent, ReadStream } from 'fs';
+import { BadRequestError, NotFoundError } from '@map-colonies/error-types';
+import { DirOperations } from '../../../../src/common/utilities';
+import { fileStreamSnap, generateRootDirSnap } from '../snapshots';
+import { streamToString } from '../utils';
```
Comment (alebinson) [inline tests/unit/storageExplorer/models/DirOperations.spec.ts:6]: imports order first "../../kuku" then "../muku"
just to be consistent with **prettier**

**Review by alebinson — CHANGES_REQUESTED** [review]
Great!!!
See my comments (not biggy I think)

File: src/common/interfaces.ts (hunk 1 above)
Comment (assafchamoy) [reply src/common/interfaces.ts:1]: fixed

File: src/common/utilities/LoggersHandler.ts (hunk 2 above)
Comment (assafchamoy) [reply src/common/utilities/LoggersHandler.ts]: fixed

File: src/common/utilities/LoggersHandler.ts (hunk 3 above)
Comment (assafchamoy) [reply src/common/utilities/LoggersHandler.ts]: fixed

File: src/storageExplorer/middlewares/getStorageExplorerMiddleware.ts (hunk 4 above)
Comment (assafchamoy) [reply src/storageExplorer/middlewares/getStorageExplorerMiddleware.ts]: fixed

File: src/storageExplorer/models/file.model.ts (hunk 5 above)
Comment (assafchamoy) [reply src/storageExplorer/models/file.model.ts]: fixed

File: tests/unit/storageExplorer/models/DirOperations.spec.ts (hunk 6 above)
Comment (assafchamoy) [reply tests/unit/storageExplorer/models/DirOperations.spec.ts:6]: fixed? my prettier was happy either way

File: src/common/utilities/dirOperations.ts [hunk 7]
```diff
@@ -0,0 +1,112 @@
+import { promises as fsPromises, Dirent, PathLike, createReadStream, statSync, existsSync } from 'fs';
+import * as Path from 'path';
+import { BadRequestError, NotFoundError, InternalServerError } from '@map-colonies/error-types';
+import { ImountDirObj, IStream } from '../interfaces';
+import { IFile, IFileMap } from '../../storageExplorer/models';
+import LoggersHandler from './LoggersHandler';
+import { encryptPath, filesArrayToMapObject } from '.';
+
+class DirOperations {
+  private readonly fileNotFoundErr = 'No such file or directory';
```
Comment (alebinson) [inline src/common/utilities/dirOperations.ts]: Just as comment, think worth (not now) to work with err_codes ex.: **fp.error.no_file_or_directory**

**Review by alebinson — APPROVED** [review]
All good, just a small comment to thikn about

File: src/common/utilities/dirOperations.ts (hunk 7 above)
Comment (assafchamoy) [reply src/common/utilities/dirOperations.ts]: OK

**Pull request merged**